src/dbnode/storage/throttler.go

```diff
@@ -0,0 +1,142 @@
+package storage
+
+import (
+	"sync"
+)
+
+// Throttler controls fair access to limited resources.
```

Comment:
I'd be specific on what "fair" here means. Specifically that you are intentionally favoring initial requests from newly blocked keys over second+ requests from existing blocked keys regardless of the order in which the requests are received. (I assume that's the intent since you are maintain a key queue rather than a request queue). Right now it takes a bit of analyzing the code to realize that is what you are going for.